### ISSUE TITLE
Got java.lang.OutOfMemoryError: unable to create new native thread when use this connector to write data in BQ using spark streaming

### ISSUE THREAD

**Issue by yirhuang** [GoogleCloudDataproc/spark-bigquery-connector#723]
```
2022-07-31T02:24:09.545954 Caused by: org.apache.spark.SparkException: Task failed while writing rows.
  at org.apache.spark.sql.execution.datasources.FileFormatWriter$.org$apache$spark$sql$execution$datasources$FileFormatWriter$$executeTask(FileFormatWriter.scala:288)
  at org.apache.spark.sql.execution.datasources.FileFormatWriter$$anonfun$write$1.apply(FileFormatWriter.scala:198)
  at org.apache.spark.sql.execution.datasources.FileFormatWriter$$anonfun$write$1.apply(FileFormatWriter.scala:197)
  at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:87)
  at org.apache.spark.scheduler.Task.run(Task.scala:109)
  at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:345)	... 3 more
Caused by: java.lang.OutOfMemoryError: unable to create new native thread
  at java.lang.Thread.start0(Native Method)
  at java.lang.Thread.start(Thread.java:717)
  at java.util.concurrent.ThreadPoolExecutor.addWorker(ThreadPoolExecutor.java:957)
  at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1367)
  at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:134)
  at com.google.cloud.hadoop.repackaged.gcs.com.google.cloud.hadoop.gcsio.BatchHelper.queue(BatchHelper.java:161)
  at com.google.cloud.hadoop.repackaged.gcs.com.google.cloud.hadoop.gcsio.GoogleCloudStorageImpl.getItemInfos(GoogleCloudStorageImpl.java:1687)
  at com.google.cloud.hadoop.repackaged.gcs.com.google.cloud.hadoop.gcsio.GoogleCloudStorageFileSystem.checkNoFilesConflictingWithDirs(GoogleCloudStorageFileSystem.java:1200)
  at com.google.cloud.hadoop.repackaged.gcs.com.google.cloud.hadoop.gcsio.GoogleCloudStorageFileSystem.create(GoogleCloudStorageFileSystem.java:251)
  at com.google.cloud.hadoop.fs.gcs.GoogleHadoopOutputStream.createChannel(GoogleHadoopOutputStream.java:78)
  at com.google.cloud.hadoop.fs.gcs.GoogleHadoopOutputStream.<init>(GoogleHadoopOutputStream.java:70)
  at com.google.cloud.hadoop.fs.gcs.GoogleHadoopSyncableOutputStream.<init>(GoogleHadoopSyncableOutputStream.java:180)
  at com.google.cloud.hadoop.fs.gcs.GoogleHadoopSyncableOutputStream.<init>(GoogleHadoopSyncableOutputStream.java:144)
  at com.google.cloud.hadoop.fs.gcs.GoogleHadoopFileSystemBase.create(GoogleHadoopFileSystemBase.java:632)
  at org.apache.hadoop.fs.FileSystem.create(FileSystem.java:1067)
  at org.apache.hadoop.fs.FileSystem.create(FileSystem.java:1048)
  at org.apache.hadoop.fs.FileSystem.create(FileSystem.java:937)
  at org.apache.parquet.hadoop.ParquetFileWriter.<init>(ParquetFileWriter.java:241)
  at org.apache.parquet.hadoop.ParquetOutputFormat.getRecordWriter(ParquetOutputFormat.java:342)
  at org.apache.parquet.hadoop.ParquetOutputFormat.getRecordWriter(ParquetOutputFormat.java:302)
  at org.apache.spark.sql.execution.datasources.parquet.ParquetOutputWriter.<init>(ParquetOutputWriter.scala:37)
  at org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat$$anon$1.newInstance(ParquetFileFormat.scala:151)
  at org.apache.spark.sql.execution.datasources.FileFormatWriter$SingleDirectoryWriteTask.newOutputWriter(FileFormatWriter.scala:370)
  at org.apache.spark.sql.execution.datasources.FileFormatWriter$SingleDirectoryWriteTask.execute(FileFormatWriter.scala:381)
  at org.apache.spark.sql.execution.datasources.FileFormatWriter$$anonfun$org$apache$spark$sql$execution$datasources$FileFormatWriter$$executeTask$3.apply(FileFormatWriter.scala:272)
  at org.apache.spark.sql.execution.datasources.FileFormatWriter$$anonfun$org$apache$spark$sql$execution$datasources$FileFormatWriter$$executeTask$3.apply(FileFormatWriter.scala:270)
  at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1415)
  at org.apache.spark.sql.execution.datasources.FileFormatWriter$.org$apache$spark$sql$execution$datasources$FileFormatWriter$$executeTask(FileFormatWriter.scala:275)	... 8 more 
```

**Comment by yirhuang** [comment]
@davidrabinowitz Thanks for your quick update.
Connection Version: spark-bigquery-with-dependencies_2.11:0.25.2
spark version: 2.4.5
java version: 1.8
Dataproc image version: not sure

Actually, I'm running the spark streaming job ingesting Kafka data to BigQuery for the duration of 1 minute.
The job ran smoothly at first but after two days, it failed with the above exception 'java.lang.OutOfMemoryError: unable to create new native thread'.

And I checked the memory & CPU used in my cluster from google console page. They are all very normal, far away from the upper limit of the instances.
Guess it may be caused by too many threads created by the spark-BQ-connector. Can you please help check if there is the possibility of any thread leak in this connector?

Thanks a lot!

**Comment by yirhuang** [comment]
![29351659618639_ pic](https://user-images.githubusercontent.com/55339063/182855132-23eb5981-a079-49b7-9752-49b88bc7eb14.jpg)
I have checked the thread dump in executors. Found number of threads starting with GAX is growing with time.

**Comment by yirhuang** [comment]
@davidrabinowitz  Any update?

**Comment by asiunov** [comment]
Same for us. The number of threads can grow to 30000 and then OOM.
bq connector 0.25.2, spark 3.2.0, scala 2.12, AdoptOpenJDK-11.0.11+9

**Comment by asiunov** [comment]
In addition to 0.25.2, I've tested 0.25.1, 0.25.0, 0.26.0 - all having the same pattern - they create Gax threads and never GC them - for the example above, I've seen all the threads from `Gax-1` to `Gax-29013` in one threaddump.

0.24.2 works fine. It does not create Gax threads, it creates `block-manager-ask-thread-pool-#` and `block-manager-storage-async-thread-pool-#` instead, but those are GCed over time, down to 0 when nothing running.

**Comment by davidrabinowitz** [comment]
Thanks fro the info - it means that the change has occurred in gax-java between version 2.13.0 and 2.17.0. We'll to narrow it down

**Comment by asnare** [comment]
Just a quick note: we're seeing this problem still with 0.27.1.

**Comment by saipraveenpn** [comment]
Facing the same issue with 0.27.0, appreciate if anyone can share a workaround

**Comment by davidrabinowitz** [comment]
@saipraveenpn Is it on read or on write? What happens if you change to the DIRECT write method?

**Comment by saipraveenpn** [comment]
@davidrabinowitz facing this in Direct write mode. However, it did go through on a retry.
Reference Scala Spark Code, please let me know if I can help with any further logs. 

`dataFrame.write
    .format("bigquery")
    .mode(SaveMode.Overwrite)
    .option("writeMethod", "direct")
    .option("parentproject", parentproject)
    .option("project", project)
    .option("dataset", dataset)
    .save(bqTable)`

**Comment by davidrabinowitz** [comment]
Hi @saipraveenpn , I have a feeling that PR #823 is not the end of it, and that it may be related to [This Apache Beam issue][(https://github.com/apache/beam/issues/21368). Once we are more confident we will release the bugfix.

In the meanwhile you can test with the `gs://spark-lib-nightly-snapshots/spark-bigquery-with-dependencies_2.12-0.0.20221219.jar` nightly snapshot (no guarantees, no SLA)

**Comment by asiunov** [comment]
Hi @davidrabinowitz , do you expect this ticket to be resolved in 0.28.0, or still waiting for this fix https://github.com/apache/beam/issues/21368 (resolved in beam repo but there is no release yet)?

**Comment by ruiyang2015** [comment]
Looks like Apache Beam has a new release: https://github.com/apache/beam/releases/tag/v2.44.0,

**Comment by dharam-cldcvr** [comment]
@davidrabinowitz Any update?
I'm also facing the issue.

Spark - 3.1.3
Scala - 2.12
Connector - 0.28.0


`User class threw exception: com.google.cloud.spark.bigquery.repackaged.com.google.inject.ProvisionException: Unable to provision, see the following errors:
1) [Guice/ErrorInCustomProvider]: OutOfMemoryError: unable to create new native thread
at BigQueryDataSourceWriterModule.provideDirectDataSourceWriterContext(BigQueryDataSourceWriterModule.java:60)
while locating BigQueryDirectDataSourceWriterContext
Learn more:
https://github.com/google/guice/wiki/ERROR_IN_CUSTOM_PROVIDER
1 error
======================
Full classname legend:
======================
BigQueryDataSourceWriterModule: “com.google.cloud.spark.bigquery.write.context.BigQueryDataSourceWriterModule”
BigQueryDirectDataSourceWriterContext: “com.google.cloud.spark.bigquery.write.context.BigQueryDirectDataSourceWriterContext”
========================
End of classname legend:
========================
at com.google.cloud.spark.bigquery.repackaged.com.google.inject.internal.InternalProvisionException.toProvisionException(InternalProvisionException.java:251)
at com.google.cloud.spark.bigquery.repackaged.com.google.inject.internal.InjectorImpl$1.get(InjectorImpl.java:1104)
at com.google.cloud.spark.bigquery.repackaged.com.google.inject.internal.InjectorImpl.getInstance(InjectorImpl.java:1139)
at com.google.cloud.spark.bigquery.write.BigQueryDataSourceWriterInsertableRelation.insert(BigQueryDataSourceWriterInsertableRelation.java:53)
at com.google.cloud.spark.bigquery.write.CreatableRelationProviderHelper.createRelation(CreatableRelationProviderHelper.java:51)
at com.google.cloud.spark.bigquery.BigQueryRelationProvider.createRelation(BigQueryRelationProvider.scala:107)
at org.apache.spark.sql.execution.datasources.SaveIntoDataSourceCommand.run(SaveIntoDataSourceCommand.scala:46)
at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult$lzycompute(commands.scala:70)
at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult(commands.scala:68)
at org.apache.spark.sql.execution.command.ExecutedCommandExec.doExecute(commands.scala:90)
at org.apache.spark.sql.execution.SparkPlan.$anonfun$execute$1(SparkPlan.scala:180)
at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeQuery$1(SparkPlan.scala:218)
at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:215)
at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:176)
at org.apache.spark.sql.execution.QueryExecution.toRdd$lzycompute(QueryExecution.scala:132)
at org.apache.spark.sql.execution.QueryExecution.toRdd(QueryExecution.scala:131)
at org.apache.spark.sql.DataFrameWriter.$anonfun$runCommand$1(DataFrameWriter.scala:989)
at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$5(SQLExecution.scala:103)
at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:163)
at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:90)
at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:775)
at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:64)
at org.apache.spark.sql.DataFrameWriter.runCommand(DataFrameWriter.scala:989)
at org.apache.spark.sql.DataFrameWriter.saveToV1Source(DataFrameWriter.scala:438)
at org.apache.spark.sql.DataFrameWriter.saveInternal(DataFrameWriter.scala:415)
at org.apache.spark.sql.DataFrameWriter.save(DataFrameWriter.scala:293)
at com.jio.bdcoe.dataingestion.flows.writer.BigQueryWriterImpl$.writeDF(BigQueryWriterImpl.scala:15)
at com.jio.bdcoe.dataingestion.flows.writer.BigQueryWriter.writeDF(BigQueryWriter.scala:8)
at com.jio.bdcoe.dataingestion.flows.writer.BigQueryWriter.writeDF$(BigQueryWriter.scala:8)
at com.jio.bdcoe.dataingestion.flows.writer.BigQueryWriterImpl$.writeDF(BigQueryWriterImpl.scala:5)
at com.jio.bdcoe.dataingestion.flows.writer.DFWriter.writeDF(DFWriter.scala:24)
at com.jio.bdcoe.dataingestion.flows.writer.DFWriter.writeDF$(DFWriter.scala:24)
at com.jio.bdcoe.dataingestion.flows.writer.BigQueryWriterImpl$.writeDF(BigQueryWriterImpl.scala:5)
at com.jio.bdcoe.dataingestion.flows.writer.DFWriter.writeDFWithContext(DFWriter.scala:32)
at com.jio.bdcoe.dataingestion.flows.writer.DFWriter.writeDFWithContext$(DFWriter.scala:26)
at com.jio.bdcoe.dataingestion.flows.writer.BigQueryWriterImpl$.writeDFWithContext(BigQueryWriterImpl.scala:5)
at com.jio.bdcoe.dataingestion.flows.IngestionHandler$.writeOutput(IngestionHandler.scala:96)
at com.jio.bdcoe.kafka.consumer.SimpleKafkaMessageProcessor.processRecords(SimpleKafkaMessageProcessor.scala:58)
at com.jio.bdcoe.kafka.consumer.BatchedMessageProcessor.doProcessBatch(BatchedMessageProcessor.scala:83)
at com.jio.bdcoe.kafka.consumer.BatchedMessageProcessor.doProcessRecords(BatchedMessageProcessor.scala:69)
at com.jio.bdcoe.kafka.consumer.KafkaConsumerWorker.call(KafkaConsumerWorker.scala:25)
at com.jio.bdcoe.kafka.consumer.OffsetAwareConsumer.$anonfun$run$1(OffsetAwareConsumer.scala:90)
at scala.util.control.Breaks.breakable(Breaks.scala:42)
at com.jio.bdcoe.kafka.consumer.OffsetAwareConsumer.run(OffsetAwareConsumer.scala:48)
at com.jio.bdcoe.kafka.consumer.KafkaMessageReceiver.$anonfun$startReceiver$1(KafkaMessageReceiver.scala:31)
at scala.collection.immutable.Range.foreach$mVc$sp(Range.scala:158)
at com.jio.bdcoe.kafka.consumer.KafkaMessageReceiver.startReceiver(KafkaMessageReceiver.scala:26)
at com.jio.bdcoe.dataingestion.flows.SimpleKafkaStreamingDriver$.main(SimpleKafkaStreamingDriver.scala:65)
at com.jio.bdcoe.dataingestion.flows.SimpleKafkaStreamingDriver.main(SimpleKafkaStreamingDriver.scala)
at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.lang.reflect.Method.invoke(Method.java:498)
at org.apache.spark.deploy.yarn.ApplicationMaster$$anon$2.run(ApplicationMaster.scala:732)
Caused by: java.lang.OutOfMemoryError: unable to create new native thread
at java.lang.Thread.start0(Native Method)
at java.lang.Thread.start(Thread.java:719)
at java.util.concurrent.ThreadPoolExecutor.addWorker(ThreadPoolExecutor.java:957)
at java.util.concurrent.ThreadPoolExecutor.ensurePrestart(ThreadPoolExecutor.java:1603)
at java.util.concurrent.ScheduledThreadPoolExecutor.delayedExecute(ScheduledThreadPoolExecutor.java:334)
at java.util.concurrent.ScheduledThreadPoolExecutor.scheduleAtFixedRate(ScheduledThreadPoolExecutor.java:573)
at com.google.cloud.spark.bigquery.repackaged.com.google.api.gax.rpc.Watchdog.start(Watchdog.java:94)
at com.google.cloud.spark.bigquery.repackaged.com.google.api.gax.rpc.Watchdog.create(Watchdog.java:82)
at com.google.cloud.spark.bigquery.repackaged.com.google.api.gax.rpc.InstantiatingWatchdogProvider.getWatchdog(InstantiatingWatchdogProvider.java:111)
at com.google.cloud.spark.bigquery.repackaged.com.google.api.gax.rpc.ClientContext.create(ClientContext.java:223)
at com.google.cloud.spark.bigquery.repackaged.com.google.cloud.bigquery.storage.v1.stub.GrpcBigQueryWriteStub.create(GrpcBigQueryWriteStub.java:132)
at com.google.cloud.spark.bigquery.repackaged.com.google.cloud.bigquery.storage.v1.stub.BigQueryWriteStubSettings.createStub(BigQueryWriteStubSettings.java:145)
at com.google.cloud.spark.bigquery.repackaged.com.google.cloud.bigquery.storage.v1.BigQueryWriteClient.<init>(BigQueryWriteClient.java:143)
at com.google.cloud.spark.bigquery.repackaged.com.google.cloud.bigquery.storage.v1.BigQueryWriteClient.create(BigQueryWriteClient.java:125)
at com.google.cloud.bigquery.connector.common.BigQueryClientFactory.createBigQueryWriteClient(BigQueryClientFactory.java:181)
at com.google.cloud.bigquery.connector.common.BigQueryClientFactory.getBigQueryWriteClient(BigQueryClientFactory.java:88)
at com.google.cloud.spark.bigquery.write.context.BigQueryDirectDataSourceWriterContext.<init>(BigQueryDirectDataSourceWriterContext.java:107)
at com.google.cloud.spark.bigquery.write.context.BigQueryDataSourceWriterModule.provideDirectDataSourceWriterContext(BigQueryDataSourceWriterModule.java:73)
at com.google.cloud.spark.bigquery.write.context.BigQueryDataSourceWriterModule$$FastClassByGuice$$3238674.GUICE$TRAMPOLINE(<generated>)
at com.google.cloud.spark.bigquery.write.context.BigQueryDataSourceWriterModule$$FastClassByGuice$$3238674.apply(<generated>)
at com.google.cloud.spark.bigquery.repackaged.com.google.inject.internal.ProviderMethod$FastClassProviderMethod.doProvision(ProviderMethod.java:260)
at com.google.cloud.spark.bigquery.repackaged.com.google.inject.internal.ProviderMethod.doProvision(ProviderMethod.java:171)
at com.google.cloud.spark.bigquery.repackaged.com.google.inject.internal.InternalProviderInstanceBindingImpl$CyclicFactory.provision(InternalProviderInstanceBindingImpl.java:185)
at com.google.cloud.spark.bigquery.repackaged.com.google.inject.internal.InternalProviderInstanceBindingImpl$CyclicFactory.get(InternalProviderInstanceBindingImpl.java:162)
at com.google.cloud.spark.bigquery.repackaged.com.google.inject.internal.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:40)
at com.google.cloud.spark.bigquery.repackaged.com.google.inject.internal.SingletonScope$1.get(SingletonScope.java:169)
at com.google.cloud.spark.bigquery.repackaged.com.google.inject.internal.InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:45)
at com.google.cloud.spark.bigquery.repackaged.com.google.inject.internal.InjectorImpl$1.get(InjectorImpl.java:1101)`

**Comment by asnare** [comment]
I'm also curious about the current status: Dataproc 2.1 is now deploying with the 0.27 version of this connector preinstalled on the classpath. I think I'm seeing this problem (massive thread counts) there, but so far our executors haven't been running long enough for this to prove fatal. (If/when that happens, I'll file a support ticket.)

**Comment by dharam-cldcvr** [comment]
My jobs are running for hours and failing all of a sudden, even with minimum workload.
Kafka -> BQ using 'direct' method.

**Comment by davidrabinowitz** [comment]
Please follow PR #951, we had to fix some of the underlying client libraries first

**Comment by kaiseu** [comment]
same issue for 0.28.1, @davidrabinowitz any ETA or target connector version on the fix? 
Spark - 3.1.3
Scala - 2.12
Connector - 0.28.1

**Comment by kaiseu** [comment]
0.24.2 works fine but seems the direct write is not stable and the [schema check](https://github.com/GoogleCloudDataproc/spark-bigquery-connector/pull/848) optimize is not there, while 0.25.0 afterwards seems all have this issue. It's a hard block especially in streaming cases, hope it can be fixed soon.

**Comment by suryasoma** [comment]
Hey, the fix for this should be available in the latest release `0.31.0` 
Thanks.
